sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ExecutionStateSampler.java

```diff
@@ -0,0 +1,418 @@
+    }
+  }
+
+  /** An {@link ExecutionState} represents the current state of an execution thread. */
+  public interface ExecutionState {
+
+    /**
+     * Activates this execution state within the {@link ExecutionStateTracker}.
+     *
+     * <p>Must only be invoked by the bundle processing thread.
+     */
+    void activate();
+
+    /**
+     * Restores the previously active execution state within the {@link ExecutionStateTracker}.
+     *
+     * <p>Must only be invoked by the bundle processing thread.
+     */
+    void deactivate();
+  }
+
+  /** Stops the execution of the state sampler. */
+  public void stop() {
+    stateSamplingThread.cancel(true);
+    try {
+      stateSamplingThread.get(5L * periodMs, TimeUnit.MILLISECONDS);
+    } catch (CancellationException e) {
+      // This was expected -- we were cancelling the thread.
+    } catch (InterruptedException | TimeoutException e) {
+      throw new RuntimeException(
+          "Failed to stop state sampling after waiting 5 sampling periods.", e);
+    } catch (ExecutionException e) {
+      throw new RuntimeException("Exception in state sampler", e);
+    }
+  }
+
+  /** Entry point for the state sampling thread. */
+  private Void stateSampler() throws Exception {
+    // Ensure the object finishes being published safely.
+    synchronized (this) {
+      if (stateSamplingThread == null) {
+        throw new IllegalStateException("Underinitialized ExecutionStateSampler instance");
+      }
+    }
+
+    long lastSampleTimeMillis = clock.getMillis();
+    long targetTimeMillis = lastSampleTimeMillis + periodMs;
+    while (!Thread.interrupted()) {
+      long currentTimeMillis = clock.getMillis();
+      long difference = targetTimeMillis - currentTimeMillis;
+      if (difference > 0) {
+        Thread.sleep(difference);
+      } else {
+        long millisSinceLastSample = currentTimeMillis - lastSampleTimeMillis;
+        synchronized (activeStateTrackers) {
+          for (ExecutionStateTracker activeTracker : activeStateTrackers) {
+            activeTracker.takeSample(currentTimeMillis, millisSinceLastSample);
+          }
+        }
+        lastSampleTimeMillis = currentTimeMillis;
+        targetTimeMillis = lastSampleTimeMillis + periodMs;
+      }
+    }
+    return null;
+  }
+
+  /** Returns a new {@link ExecutionStateTracker} associated with this state sampler. */
+  public ExecutionStateTracker create() {
+    return new ExecutionStateTracker();
+  }
+
+  /** Tracks the current state of a single execution thread. */
+  public class ExecutionStateTracker implements BundleProgressReporter {
+
+    // The set of execution states that this tracker is responsible for. Effectively
+    // final after the first execution of start().
+    private final List<ExecutionStateImpl> executionStates;
+    // Read by multiple threads, written by the bundle processing thread lazily.
+    private final AtomicReference<@Nullable String> processBundleId;
```

Comment:
just to make sure I'm understanding this, the semantics of lazySet is basically a store-store barrier, so other threads may not observe the `set` for some amount of time.  We're ok with this since the sampling doesn't need to be exact?

Reply:
That is exactly right.